### PR TITLE
WIP: pandas >= 2.0 support

### PR DESCRIPTION
**Pull request by wasade**
Please complete the following checklist:

* [x] I have read the guidelines in [CONTRIBUTING.md](https://github.com/biocore/scikit-bio/blob/master/.github/CONTRIBUTING.md).

* [x] I have documented all public-facing changes in [CHANGELOG.md](https://github.com/biocore/scikit-bio/blob/master/CHANGELOG.md).

* [ ] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/biocore/scikit-bio/blob/master/LICENSE.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied. **It is your responsibility to disclose code, documentation, or other content derived from external source(s).** If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [x] **This pull request does not include code, documentation, or other content derived from external source(s).**

**Note:** [REVIEWING.md](https://github.com/biocore/scikit-bio/blob/master/REVIEWING.md) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

### CODE REVIEW

**Comment by wasade** [comment]
@ebolyen, this PR is going to fail due to inhomogenous arrays. I'm not familiar with the affected parts of the codebase, so I'm at the moment unsure of a resolution. For context, one of the failures (with its output printed) is below, and I'm quite puzzled over what is correct as the content of `indexable` is jagged. Do you have any thoughts here? 

```
    def _assert_bool_vector_right_size(self, indexable, axis):
        if isinstance(indexable, np.ndarray):
            pass
        elif hasattr(indexable, '__iter__'):
            print(indexable)
>           indexable = np.asarray(list(indexable))
E           ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (3,) + inhomogeneous part.

skbio/alignment/_indexing.py:117: ValueError
---------------------------------------------- Captured stdout call ----------------------------------------------
([False, True, False, True], 'x', 2)
```

**Comment by ebolyen** [comment]
Well it appears the `unthinkable` index has finally met it's match. I think the sanest thing to do would be to drop support and document it.

The tests for array indexing in the seq+msa objects were quite heavily fit to whatever pandas could possibly do. Since the 2.0 version probably let them eject some absurd behaviors, we may as well follow.

File: skbio/alignment/tests/test_tabular_msa.py [hunk 1]
```diff
@@ -1360,41 +1360,6 @@ def test_multiindex_fancy_indexing_incomplete_label(self):
...
-        msa = TabularMSA([a, b, c, d], metadata={'x': 'y'},
-                         positional_metadata={'c': ['a', 'b', 'c', 'd']},
-                         index=[('a', 'x', 0), ('a', 'x', 1), ('a', 'y', 2),
-                                ('b', 'x', 0)])
-        # Pandas will KeyError when the intersection is empty
-        # change appears to have happened in:
-        # https://github.com/pandas-dev/pandas/pull/42245
-        # but this was not bisected to confirm
-        with self.assertRaises(KeyError):
-            self.get(msa, (([False, True, False, True], 'x', 2), Ellipsis))
```
Comment (wasade) [inline skbio/alignment/tests/test_tabular_msa.py:1396]: @ebolyen, you okay with this?

Comment (ebolyen) [reply skbio/alignment/tests/test_tabular_msa.py:1396]: Sorry, did not see this! Yeah that's totally fine, these were observational tests more than good features.

Comment (wasade) [reply skbio/alignment/tests/test_tabular_msa.py:1396]: Wonderful, thank you @ebolyen!

**Pull request merged**